### PR TITLE
Civ 6069

### PR DESCRIPTION
**Pull request by maciejmodzelewski**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/CIV-6069

### Change description ###
A new custom exception added and used instead of RuntimeException

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/befta/dse/ccd/DataLoaderToDefinitionStore.java [hunk 1]
```diff
@@ -315,24 +316,35 @@ protected void importDefinitionsAt(String definitionsPath) {
...
-                    if (!shouldTolerateDataSetupFailure()) {
-                        throw new RuntimeException(e);
+                } catch (ImportException e) {
+                    logger.error(message, fileName, e);
+                    if (!shouldTolerateDataSetupFailure(e)) {
+                        throw e;
                     }
+                } catch (RuntimeException e) {
+                    logger.error(message, fileName, e);
+                        throw e;
```
Comment (filippo-possenti-version1) [inline src/main/java/uk/gov/hmcts/befta/dse/ccd/DataLoaderToDefinitionStore.java:333]: Minor: indentation

File: src/main/java/uk/gov/hmcts/befta/dse/ccd/DataLoaderToDefinitionStore.java [hunk 2]
```diff
@@ -315,24 +316,35 @@ protected void importDefinitionsAt(String definitionsPath) {
...
+                    throw new RuntimeException(e);
                 }
             }
         } finally {
             FileUtils.deleteDirectory(TEMPORARY_DEFINITION_FOLDER);
         }
     }
 
+    protected boolean shouldTolerateDataSetupFailure(Throwable e) {
+        return this.shouldTolerateDataSetupFailure();
```
Comment (filippo-possenti-version1) [inline src/main/java/uk/gov/hmcts/befta/dse/ccd/DataLoaderToDefinitionStore.java:345]: Minor: "this" can be omitted

**Review by filippo-possenti-version1 — APPROVED** [review]

File: src/main/java/uk/gov/hmcts/befta/dse/ccd/DataLoaderToDefinitionStore.java (hunk 1 above)
Comment (maciejmodzelewski) [reply src/main/java/uk/gov/hmcts/befta/dse/ccd/DataLoaderToDefinitionStore.java:333]: Fixed

File: src/main/java/uk/gov/hmcts/befta/dse/ccd/DataLoaderToDefinitionStore.java (hunk 2 above)
Comment (maciejmodzelewski) [reply src/main/java/uk/gov/hmcts/befta/dse/ccd/DataLoaderToDefinitionStore.java:345]: "this" removed

File: src/main/java/uk/gov/hmcts/befta/dse/ccd/DataLoaderToDefinitionStore.java [hunk 3]
```diff
@@ -315,24 +316,35 @@ protected void importDefinitionsAt(String definitionsPath) {
...
                     logger.info("\n\nImporting {}...", fileName);
                     importDefinition(fileName);
                     logger.info("\nImported {}.\n\n", fileName);
-                } catch (Exception e) {
-                    logger.error("Couldn't import {} - Exception: {}.\n\n", fileName, e);
-                    if (!shouldTolerateDataSetupFailure()) {
-                        throw new RuntimeException(e);
+                } catch (ImportException e) {
+                    logger.error(message, fileName, e);
+                    if (!shouldTolerateDataSetupFailure(e)) {
```
Comment (rajaram-nagarajan) [inline src/main/java/uk/gov/hmcts/befta/dse/ccd/DataLoaderToDefinitionStore.java:328]: we can avoid this new method with throwable parameter when we actually don't have any logic

File: src/main/java/uk/gov/hmcts/befta/dse/ccd/DataLoaderToDefinitionStore.java [hunk 4]
```diff
@@ -315,24 +316,35 @@ protected void importDefinitionsAt(String definitionsPath) {
...
             for (String fileName : definitionFileResources) {
                 try {
                     logger.info("\n\nImporting {}...", fileName);
                     importDefinition(fileName);
                     logger.info("\nImported {}.\n\n", fileName);
-                } catch (Exception e) {
-                    logger.error("Couldn't import {} - Exception: {}.\n\n", fileName, e);
-                    if (!shouldTolerateDataSetupFailure()) {
-                        throw new RuntimeException(e);
+                } catch (ImportException e) {
```
Comment (rajaram-nagarajan) [inline src/main/java/uk/gov/hmcts/befta/dse/ccd/DataLoaderToDefinitionStore.java:326]: We can add new test cases to cover these changes

File: src/main/java/uk/gov/hmcts/befta/dse/ccd/DataLoaderToDefinitionStore.java (hunk 3 above)
Comment (filippo-possenti-version1) [reply src/main/java/uk/gov/hmcts/befta/dse/ccd/DataLoaderToDefinitionStore.java:328]: We have that logic in civils. That's why we made this change.

Basically we noticed that while importing an HTTP-504 can be thrown but in that case the import succeeds (it's just too many imports being processed) so we want to ignore the HTTP-504. What we don't want to ignore is all other errors.

As a result we added a `shouldTolerateDataSetupFailure` method override that accepts one argument being our exception, so we can check the actual HTTP status code.

Adding this overload is the whole point of our change.

File: src/main/java/uk/gov/hmcts/befta/dse/ccd/DataLoaderToDefinitionStore.java (hunk 4 above)
Comment (filippo-possenti-version1) [reply src/main/java/uk/gov/hmcts/befta/dse/ccd/DataLoaderToDefinitionStore.java:326]: The whole `importDefinitionsAt` (and `importDefinitions`) method is not covered by tests. It's not so much this change that would need to be covered but the whole method.

Basically to me that's technical debt that I don't think should be handled as part of this ticket.

Comment (rajaram-nagarajan) [reply src/main/java/uk/gov/hmcts/befta/dse/ccd/DataLoaderToDefinitionStore.java:326]: agreed, the tests would be covered under different ticket.

File: src/main/java/uk/gov/hmcts/befta/dse/ccd/DataLoaderToDefinitionStore.java (hunk 3 above)
Comment (rajaram-nagarajan) [reply src/main/java/uk/gov/hmcts/befta/dse/ccd/DataLoaderToDefinitionStore.java:328]: along with 504 (which is a common behaviour for all service teams), we have to consider other 2xx scenarios as well. So, I am approving this for now to resolve your problem. We will plan to tidy up the http response code handling in the future.

**Review by rajaram-nagarajan — APPROVED** [review]
LGTM

**Pull request merged**